internal/integration/unified/entity.go

```diff
@@ -579,135 +601,135 @@ func getKmsCredential(kmsDocument bson.Raw, credentialName string, envVar string
 	var ok bool
 	var doc bson.Raw
 	if doc, ok = credentialVal.DocumentOK(); !ok {
-		return "", fmt.Errorf("expected String or Document for %v, got: %v", credentialName, credentialVal)
+		return nil, fmt.Errorf("expected String or Document for %v, got: %v", credentialName, credentialVal)
 	}
 
-	placeholderDoc := bsoncore.NewDocumentBuilder().AppendInt32("$$placeholder", 1).Build()
-
 	// Check if document is a placeholder.
 	if !bytes.Equal(doc, placeholderDoc) {
-		return "", fmt.Errorf("unexpected non-empty document for %v: %v", credentialName, doc)
+		return nil, fmt.Errorf("unexpected non-empty document for %v: %v", credentialName, doc)
 	}
+
 	if envVar == "" {
 		return defaultValue, nil
 	}
-	if os.Getenv(envVar) == "" {
-		if defaultValue != "" {
-			return defaultValue, nil
-		}
-		return "", fmt.Errorf("unable to get environment value for %v. Please set the CSFLE environment variable: %v", credentialName, envVar)
+	if value := os.Getenv(envVar); value != "" {
+		return value, nil
 	}
-	return os.Getenv(envVar), nil
-
+	if defaultValue != "" {
+		return defaultValue, nil
+	}
+	return nil, fmt.Errorf("unable to get environment value for %v. Please set the CSFLE environment variable: %v", credentialName, envVar)
 }
 
-func (em *EntityMap) addClientEncryptionEntity(entityOptions *entityOptions) error {
-	// Construct KMS providers.
-	kmsProviders := make(map[string]map[string]any)
-	ceo := entityOptions.ClientEncryptionOpts
-	tlsconf := make(map[string]*tls.Config)
-	if aws, ok := ceo.KmsProviders["aws"]; ok {
-		kmsProviders["aws"] = make(map[string]any)
+func getKmsProvider(key string, opt bson.Raw) (map[string]any, error) {
+	provider := make(map[string]any)
+	switch key {
+	case "aws":
+		accessKeyID := "FLE_AWS_KEY"
+		secretAccessKey := "FLE_AWS_SECRET"
 
-		awsSessionToken, err := getKmsCredential(aws, "sessionToken", "CSFLE_AWS_TEMP_SESSION_TOKEN", "")
+		// replace with temporary access, if sessionToken placeholder exists
+		v, err := getKmsCredential(opt, "sessionToken", "", "$$placeholder")
 		if err != nil {
-			return err
+			return nil, err
+		}
+		if v == "$$placeholder" {
+			provider["sessionToken"] = os.Getenv("CSFLE_AWS_TEMP_SESSION_TOKEN")
+			accessKeyID = "CSFLE_AWS_TEMP_ACCESS_KEY_ID"
+			secretAccessKey = "CSFLE_AWS_TEMP_SECRET_ACCESS_KEY"
+		} else if v != nil {
+			provider["sessionToken"] = v
 		}
-		if awsSessionToken != "" {
-			// Get temporary AWS credentials.
-			kmsProviders["aws"]["sessionToken"] = awsSessionToken
-			awsAccessKeyID, err := getKmsCredential(aws, "accessKeyId", "CSFLE_AWS_TEMP_ACCESS_KEY_ID", "")
-			if err != nil {
-				return err
-			}
-			if awsAccessKeyID != "" {
-				kmsProviders["aws"]["accessKeyId"] = awsAccessKeyID
-			}
 
-			awsSecretAccessKey, err := getKmsCredential(aws, "secretAccessKey", "CSFLE_AWS_TEMP_SECRET_ACCESS_KEY", "")
+		for _, e := range []struct {
+			key    string
+			envVar string
+		}{
+			{"accessKeyId", accessKeyID},
+			{"secretAccessKey", secretAccessKey},
```

Comment:
[nit] Suggest keying the field values for readability and breaking parameterization.